File: clojure/luhn/src/luhn.clj
(ns luhn
  (:require [clojure.string :as str]))

(defn digits [n]
  (map #(- (int %) (int \0)) (str/replace (str n) #"\D" "")))

(defn- wrap [n] (if (>= n 10) (- n 9) n))

(defn checksum[n]
  (let [ds (partition 2 2 [0] (reverse (digits n)))]
    (rem (apply + (map #(+ (wrap (first %)) (wrap (* 2 (second %))))
                       ds))
         10)))

(defn valid? [n]
  (zero? (checksum n)))

(defn add-check-digit [n]
  (let [num (* n 10)]
    (first (for [check-digit (range 10)
                 :when (valid? (+ num check-digit))]
             (+ num check-digit)))))

File: go/luhn/luhn.go
package luhn

import (
	"strconv"
	"strings"
)

func Valid(s string) bool {
	if strings.TrimSpace(s) == "" {
		return false
	}
	return checksum(toUint64(s)) == 0
}

func AddCheck(s string) string {
	return ""
}

func toUint64(s string) uint64 {
	i, _ := strconv.ParseUint(strings.Replace(s, " ", "", -1), 10, 64)
	return i
}

func checksum(i uint64) int {
	csum := 0
	for i != 0 {
		first := int(i % 10)
		second := int((i / 10) % 10)
		i /= 100
		csum += wrap(first) + wrap(second*2)
	}
	return csum % 10
}

func wrap(i int) int {
	if i >= 10 {
		return i - 9
	}
	return i
}
